app/src/main/kotlin/io/intrepid/skotlinton/base/BaseMvpActivity.kt

```diff
@@ -1,67 +1,67 @@
 package io.intrepid.skotlinton.base
 
-import android.content.Intent
+import android.arch.lifecycle.ViewModel
+import android.arch.lifecycle.ViewModelProvider
+import android.arch.lifecycle.ViewModelProviders
 import android.os.Bundle
 import android.support.annotation.CallSuper
+import io.reactivex.disposables.CompositeDisposable
 
 /**
  * Base class for activities that will have some business logic instead of just hosting a fragment.
  * If the activity is only going to act as a container for a fragment, use {@link BaseFragmentActivity}
  * instead
  */
-abstract class BaseMvpActivity<out P : BaseContract.Presenter> : BaseActivity(), BaseContract.View {
+abstract class BaseMvpActivity<VM : BaseViewModel> : BaseActivity() {
 
-    protected val presenter: P by lazy(LazyThreadSafetyMode.NONE) {
-        val configuration = skotlintonApplication.getPresenterConfiguration()
-        createPresenter(configuration)
+    protected val onPauseDisposable = CompositeDisposable()
+    protected val onStopDisposable = CompositeDisposable()
+    protected val onDestroyDisposable = CompositeDisposable()
+
+    @Suppress("UNCHECKED_CAST")
+    protected val viewModel: VM by lazy(LazyThreadSafetyMode.NONE) {
+        val factory = object : ViewModelProvider.Factory {
+            override fun <T : ViewModel> create(modelClass: Class<T>): T {
+                val configuration = skotlintonApplication.getViewModelConfiguration()
+                return createViewModel(configuration) as T
+            }
+        }
+        ViewModelProviders.of(this, factory).get(viewModelClass) as VM
     }
 
+    abstract val viewModelClass: Class<out ViewModel>
```

Comment:
we can't infer, thanks to Java type erasure

Reply:
Actually, since the factory never uses this parameter, we could probably get away with dropping this property and just passing `ViewModel::class.java` to `ViewModelProvider.get()`

Reply:
I'm not sure if I understand this part. The `ViewModelProvider.get()` source code does indicate that the parameter is used

Reply:
So what I had been thinking here is that, ultimately what `ViewModelProvider` does is take this class and forward it to the factory we provide, which just calls `createViewModel()` and doesn't care what the class is. But I was ignoring the fact that `ViewModelProvider` does some caching based on the class, which would break if we always passed `ViewModel::class.java`. So yes we should leave this as-is.